File: nexxT/src/Executor.cpp
/*
 * SPDX-License-Identifier: Apache-2.0
 * Copyright (C) 2020 ifm electronic gmbh
 *
 * THE PROGRAM IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND.
 */

#include "Executor.hpp"
#include "FilterEnvironment.hpp"
#include "InputPortInterface.hpp"
#include "OutputPortInterface.hpp"
#include "Logger.hpp"

#include <QtCore/QThread>
#include <QtCore/Qt>
#include <QtCore/QMutex>
#include <QtCore/QMutexLocker>
#include <QtCore/QSemaphore>
#include <QtCore/QTimer>

#include <vector>
#include <set>
#include <chrono>

/* maximum number of events processed before step function returns */
#define MAX_EVENTS_PER_STEP (32)
/* after this time has elapsed, the step function will return to avoid unresponsiveness */
#define STEP_DEADLINE (100ms)

using namespace nexxT;

namespace nexxT
{
    struct ExecutorD
    {
        struct ReceiveEvent
        {
            SharedInputPortPtr inputPort;
            SharedDataSamplePtr dataSample;
            QSemaphore *semaphore;
        };

        static const int MAX_LOOPS_FINALIZE = 5;

        QMutex pendingReceivesMutex;
        std::vector<ReceiveEvent> pendingReceives;
        std::set<Filter *> blockedFilters;
        bool stopped;
        // in contrast to the python reference implementation, the pressure on the QT event loop is reduced in the C++
        // implementation by omitting unnecessary notify calls through the event loop. We count the number of pending
        // notify calls and omit the call if there is already a pending call in the event loop.
        int32_t numNotifiesInQueue;
    };
};

Executor::Executor(QThread *qthread) :
    QObject(),
    d(new ExecutorD())
{
    moveToThread(qthread);
    QObject::connect(this, SIGNAL(notify()), this, SLOT(notifyInThread()), Qt::QueuedConnection);
    d->numNotifiesInQueue = 0;
}

Executor::~Executor()
{
    delete d;
}

void Executor::registerPendingRcvSync(const SharedInputPortPtr &inputPort,
                                      const SharedDataSamplePtr &dataSample)
{
    if( !d->stopped )
    {
        ExecutorD::ReceiveEvent ev{inputPort, dataSample, 0};
        {
            QMutexLocker locker(&d->pendingReceivesMutex);
            //NEXXT_LOG_INFO(QString("add to executor queue %1/%2 (sync)").
            //    arg(inputPort->environment()->getFullQualifiedName()).
            //    arg(inputPort->name()));
            d->pendingReceives.push_back(ev);
        }
        notifyInThread();
    }
}

void Executor::registerPendingRcvAsync(const SharedInputPortPtr &inputPort,
                                       const SharedDataSamplePtr &dataSample,
                                       QSemaphore *semaphore)
{
    if( !d->stopped )
    {
        ExecutorD::ReceiveEvent ev{inputPort, dataSample, semaphore};
        {
            QMutexLocker locker(&d->pendingReceivesMutex);
            //NEXXT_LOG_INFO(QString("add to executor queue %1/%2 (async)").
            //   arg(inputPort->environment()->getFullQualifiedName()).arg(inputPort->name()));
            d->pendingReceives.push_back(ev);
        }
        emit notify();
    }
}

void Executor::notifyInThread()
{
    if( QThread::currentThread() != thread() )
    {
        NEXXT_LOG_ERROR("Executor::notifyInThread: Unexpected thread!");
    }
    if(d->numNotifiesInQueue == 0)
    {
        //NEXXT_LOG_INFO(QString("[%1] calling step via QT event.").arg(QThread::currentThread()->objectName()));
        d->numNotifiesInQueue++;
        QTimer::singleShot(0, this, SLOT(multiStep()));
    }
}

void Executor::multiStep()
{
    using namespace std::literals::chrono_literals;
    std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
    std::chrono::steady_clock::time_point curr = begin;
    uint32_t maxEvents = MAX_EVENTS_PER_STEP;
    d->numNotifiesInQueue--;
    if(d->numNotifiesInQueue < 0)
    {
        NEXXT_LOG_ERROR("Unexpected numNotifiesInQueue!");
    }
    while( (!d->stopped) && (maxEvents > 0) && (curr - begin < STEP_DEADLINE) )
    {
        if(!step())
        {
            /* no more events to process */
            return;
        } else
        {
            maxEvents--;
        }
        curr = std::chrono::steady_clock::now();
    }
    /* the last step() function result was true, so we aborted early, register another multiStep call */
    notifyInThread();
}

struct StepFunctionHelper
{
    const SharedFilterPtr &fromFilter;
    ExecutorD *d;
    bool &res;

    StepFunctionHelper(const SharedFilterPtr &_fromFilter, ExecutorD *_d, bool &_res) :
        fromFilter(_fromFilter),
        d(_d),
        res(_res)
    {
        if( fromFilter.get() != 0 )
        {
            //NEXXT_LOG_INFO(QString("[%1] Entering Executor::step, blocking filter %2").
            //    arg(QThread::currentThread()->objectName()).
            //    arg(fromFilter->environment()->getFullQualifiedName()));
            d->blockedFilters.insert(fromFilter.get());
        } else
        {
            //NEXXT_LOG_INFO(QString("[%1] Entering Executor::step without blocking").
            //    arg(QThread::currentThread()->objectName()));
        }
    }
    ~StepFunctionHelper()
    {
        if( !res )
        {
            d->pendingReceivesMutex.unlock();
        }
        if( fromFilter.get() != 0 )
        {
            //NEXXT_LOG_INFO(QString("[%1] Unblocking filter %2").
            //    arg(QThread::currentThread()->objectName()).
            //    arg(fromFilter->environment()->getFullQualifiedName()));
            d->blockedFilters.erase(fromFilter.get());
        }
        //NEXXT_LOG_INFO(QString("[%1] Leaving Executor::step").
        //    arg(QThread::currentThread()->objectName()));
    }
};

bool Executor::step(const SharedFilterPtr &fromFilter)
{
    bool res = false;
    if( !d->stopped )
    {
        StepFunctionHelper helper(fromFilter, d, res);
        d->pendingReceivesMutex.lock();
        for(auto it=d->pendingReceives.begin(); it != d->pendingReceives.end(); it++)
        {
            if( d->blockedFilters.empty() ||
                d->blockedFilters.count(it->inputPort->environment()->getPlugin().get()) == 0 )
            {
                ExecutorD::ReceiveEvent ev(*it);
                d->pendingReceives.erase(it);
                /* it is invalid from here on */
                d->pendingReceivesMutex.unlock();
                res = true;
                if( !ev.semaphore )
                {
                    ev.inputPort->receiveSync(ev.dataSample);
                } else
                {
                    ev.inputPort->receiveAsync(ev.dataSample, ev.semaphore);
                }
                break;
            }
        }
    }
    return res;
}

void Executor::finalize()
{
    std::multiset<InputPortInterface *> numCalled;
    bool changed = true;
    while(changed)
    {
        changed = false;
        d->pendingReceivesMutex.lock();
        for(auto it=d->pendingReceives.begin(); it != d->pendingReceives.end(); it++)
        {
            bool cond1 = d->blockedFilters.count(it->inputPort->environment()->getPlugin().get()) == 0;
            bool cond2 = numCalled.count(it->inputPort.get()) < d->MAX_LOOPS_FINALIZE;
            if(cond1 && cond2)
            {
                ExecutorD::ReceiveEvent ev(*it);
                d->pendingReceives.erase(it);
                /* it is invalid from here on */
                d->pendingReceivesMutex.unlock();
                changed = true;
                numCalled.insert(ev.inputPort.get());
                if( !ev.semaphore )
                {
                    ev.inputPort->receiveSync(ev.dataSample);
                } else
                {
                    ev.inputPort->receiveAsync(ev.dataSample, ev.semaphore);
                }
                d->pendingReceivesMutex.lock();
                break;
            }
        }
        d->pendingReceivesMutex.unlock();
    }
}

void Executor::clear()
{
    d->stopped = true;
    d->pendingReceives.clear();
    d->blockedFilters.clear();
}

SharedExecutorPtr Executor::make_shared(Executor *executor)
{
    return SharedExecutorPtr(executor);
}

File: nexxT/core/Executor.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (C) 2020 ifm electronic gmbh
#
# THE PROGRAM IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND.
#

"""
This module defines the class Executor
"""

import logging
from PySide2.QtCore import QObject, Signal, QThread, Qt, QMutex, QTimer
import nexxT
from nexxT.core.Utils import handleException

logger = logging.getLogger(__name__)

if nexxT.useCImpl:

    import cnexxT

    def Executor(*args): # pylint: disable=invalid-name
        """
        This is a factory function for the C version of the Executor class
        """
        res = cnexxT.nexxT.Executor.make_shared(cnexxT.nexxT.Executor(*args))
        return res

else:

    class Executor(QObject):
        """
        Each nexxT thread has an executor for executing tasks (i.e. notifying filters about changes in the ports).
        """

        notify = Signal()
        MAX_LOOPS_FINALIZE = 5

        def __init__(self, qthread):
            """
            Construtor

            :param qthread: the QThread instance this object shall be moved to
            """
            super().__init__()
            self._pendingReceivesMutex = QMutex()
            self._pendingReceives = []
            self._blockedFilters = set()
            self._stopped = False
            self.moveToThread(qthread)
            self.notify.connect(self.notifyInThread, Qt.QueuedConnection)

        def registerPendingRcvSync(self, inputPort, dataSample):
            """
            Register a pending synchronous (i.e. originated from the same thread) receive event.

            :param inputPort: The InputPort instance which shall be notified
            :param dataSample: The DataSample instance to be delivered
            """
            self._registerPendingRcvSync(inputPort, dataSample)

        @handleException
        def _registerPendingRcvSync(self, inputPort, dataSample):
            if not self._stopped:
                self._pendingReceivesMutex.lock()
                self._pendingReceives.append((inputPort, dataSample, None))
                self._pendingReceivesMutex.unlock()
                QTimer.singleShot(0, self.step)

        def registerPendingRcvAsync(self, inputPort, dataSample, semaphore):
            """
            Register a pending asynchronous (i.e. originated from another thread through a PortToPortConnection
            instacne) receive event.

            :param inputPort: The InputPort instance which shall be notified
            :param dataSample: The DataSample instance to be delivered
            :param semaphore: The QSemaphore instance of the corresponding InterThreadConnection instance
            """
            self._registerPendingRcvAsync(inputPort, dataSample, semaphore)

        @handleException
        def _registerPendingRcvAsync(self, inputPort, dataSample, semaphore):
            assert inputPort.thread() == self.thread()
            if not self._stopped:
                self._pendingReceivesMutex.lock()
                self._pendingReceives.append((inputPort, dataSample, semaphore))
                self._pendingReceivesMutex.unlock()
                self.notify.emit()

        def notifyInThread(self):
            """
            Slot called during _registerPendingRcvAsync, starts a single shot timer for the step function.
            """
            self._notifyInThread()

        @handleException
        def _notifyInThread(self):
            QTimer.singleShot(0, self.step)

        def step(self, fromFilter=None):
            """
            This function process one of the pending events.

            :param fromFilter: An optional filter instance which will be blocked for further processing until the
                function returns
            :return: True if an event was processed, False otherwise.
            """
            return self._step(fromFilter)

        @handleException
        def _step(self, fromFilter):
            assert QThread.currentThread() == self.thread()
            res = False
            if fromFilter is not None:
                self._blockedFilters.add(fromFilter)
            try:
                if not self._stopped:
                    self._pendingReceivesMutex.lock()
                    for idx, (inputPort, dataSample, semaphore) in enumerate(self._pendingReceives):
                        if inputPort.environment().getPlugin() not in self._blockedFilters:
                            self._pendingReceives.pop(idx)
                            self._pendingReceivesMutex.unlock()
                            res = True
                            if semaphore is None:
                                inputPort.receiveSync(dataSample)
                            else:
                                inputPort.receiveAsync(dataSample, semaphore)
                            # only process one sample
                            break
            finally:
                if not res:
                    self._pendingReceivesMutex.unlock()
                if fromFilter is not None:
                    self._blockedFilters.remove(fromFilter)
            return res

        def finalize(self):
            """
            This function processes the queue before the thread is stopped. In case of infinite recursions, an early
            stop criterion is applied.
            """
            logger.internal("starting finalize (%s)", self.thread())
            numCalled = {}
            changed = True
            while changed:
                changed = False
                self._pendingReceivesMutex.lock()
                for idx, (inputPort, dataSample, semaphore) in enumerate(self._pendingReceives):
                    if (inputPort.environment().getPlugin() not in self._blockedFilters and
                            numCalled.get(inputPort, 0) < self.MAX_LOOPS_FINALIZE):
                        self._pendingReceives.pop(idx)
                        self._pendingReceivesMutex.unlock()
                        numCalled[inputPort] = numCalled.get(inputPort, 0) + 1
                        changed = True
                        if semaphore is None:
                            inputPort.receiveSync(dataSample)
                        else:
                            inputPort.receiveAsync(dataSample, semaphore)
                        self._pendingReceivesMutex.lock()
                        # only one sample, note that _pendingReceives is changed inside the loop!
                        break
                self._pendingReceivesMutex.unlock()

        def clear(self):
            """
            Called after processing is stopped.
            """
            self._stopped = True
            self._pendingReceives = []
            self._blockedFilters = []
